src/main/java/com/pinterest/secor/uploader/FormatConvertingUploadManager.java

```diff
@@ -0,0 +1,154 @@
+ */
+package com.pinterest.secor.uploader;
+
+import com.pinterest.secor.common.LogFilePath;
+import com.pinterest.secor.common.SecorConfig;
+import com.pinterest.secor.io.FileReader;
+import com.pinterest.secor.io.FileWriter;
+import com.pinterest.secor.io.KeyValue;
+import com.pinterest.secor.uploader.UploadManager;
+import com.pinterest.secor.util.CompressionUtil;
+import com.pinterest.secor.util.FileUtil;
+import com.pinterest.secor.util.IdUtil;
+import com.pinterest.secor.util.ReflectionUtil;
+
+import java.io.File;
+
+import org.apache.hadoop.io.compress.CompressionCodec;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages uploads using the configured upload manager
+ * Converts the local files to the specified format before upload
+ *
+ * @author Jason Butterfield (jason.butterfield@outreach.io)
+ */
+public class FormatConvertingUploadManager extends UploadManager {
+    private static final Logger LOG = LoggerFactory.getLogger(FormatConvertingUploadManager.class);
+
+    private UploadManager mUploadManager;
+
+    public FormatConvertingUploadManager(SecorConfig config) throws Exception {
+        super(config);
+        mUploadManager = createUploadManager();
+    }
+
+    @Override
+    public TempFileUploadHandle<Handle<?>> upload(LogFilePath localPath) throws Exception {
+        // convert the file from the internal format to the external format
+        System.out.println("localPath: " + localPath);
+        System.out.println("mUploadManager: " + mUploadManager);
+        LogFilePath convertedFilePath = convertFile(localPath);
+
+        // The TempFileUploadHandle will delete the temp file after it resolves
+        return new TempFileUploadHandle(mUploadManager.upload(convertedFilePath), convertedFilePath);
+    }
+
+    /**
+     * This method converts a file 
+     * Exposed for testing only
+     */
+    public LogFilePath convertFile(LogFilePath srcPath) throws Exception {
+        FileReader reader = null;
+        FileWriter writer = null;
+        int copiedMessages = 0;
+
+        String localConvertedPrefix = mConfig.getLocalPath() + '/' +
+            IdUtil.getLocalMessageDir() + "/convertedForUpload";
+        LogFilePath convertedFilePath = srcPath.withPrefix(localConvertedPrefix);
+
+        try {
+            CompressionCodec codec = null;
+            String extension = "";
+            if (mConfig.getCompressionCodec() != null && !mConfig.getCompressionCodec().isEmpty()) {
+                codec = CompressionUtil.createCompressionCodec(mConfig.getCompressionCodec());
+                extension = codec.getDefaultExtension();
+            }
+
+            reader = createReader(srcPath, codec);
+            writer = createWriter(convertedFilePath, codec);
+
+            KeyValue keyVal;
+            while ((keyVal = reader.next()) != null) {
+              writer.write(keyVal);
+              copiedMessages++;
+            }
+        } finally {
+            if (reader != null) {
```

Comment:
👍 Good catch on finally

Reply:
I leaked a fucking selenium browser once forgetting this in ruby hahahaha

Reply:
It was like a 200-450MB leak every time it happened!!!

Reply:
Ouch!